README.md

### A low-cost Flutter screen adaptation solution（一个极低成本的 Flutter 屏幕适配方案）
### 100% 还原 UI，只需要按照设计图写的宽高写即可
# 先看图片，设置的标准宽度是 360
###  iPhone 8 ------------------------------- iPhone 11 :
<div >
   <img style="vertical-align: top;" src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/iPhone%208.png?raw=true" width="40%" div align=top>
   <img style="vertical-align: top;" src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/iPhone%2011.png?raw=true" width="40%" >
</div>

###  iPhone 12 pro max  ---------------------  ipad air :
<div >
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/iPhone%2012.png?raw=true" width="40%" div align=top>
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/iPad%20Air.png?raw=true" width="40%" div align=top >
</div>

## android 图
### 768x1280-320dpi----------10801920-420dpi----------1440x2560-560dpi
<div >
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/768*1280-320dpi.png?raw=true" width="30%" div align=top>
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/1080*1920-420dpi.png?raw=true" width="30%" div align=top >
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/1440*2560-560dpi.png?raw=true" width="30%" div align=top >
</div>

## web 图
随着拉长屏幕 ，慢慢的 宽度会大于高度，当大于的时候 ，会以 高度 为 基准。
<div >
   <img src="https://github.com/niezhiyang/flutter_autosize_screen/blob/master/art/web.gif?raw=true">
</div>

## 使用
### 引用
```dart
flutter_autosize_screen: ^{LAST_VERSION}
```
### 初始化
1. 在main方法的第一行就初始化，下面的基准一般以宽度为基准，直接写Ui设计图的宽度尺寸，如果是横屏的状态的 下面的 360 就是以高度为基准
```dart
void main() {
  // 设置基准 或者以苹果6，375 为标准, isAutoTextSize 表示文字是否随着系统的字体大小更改而更改，默认是 true
  AutoSizeUtil.setStandard(360,isAutoTextSize: true);

  // 使用 runAutoApp 来代替 runApp
  // import 'package:flutter_autosize_screen/binding.dart';
  runAutoApp(const MyApp());

}

```
2. 替换根 MaterialApp 的 MediaQuery
```dart
MaterialApp(
      title: 'Autosize Demo',
      /// 替换，这样可以在以后 使用 MediaQuery.of(context) 得到 正确的Size
      builder: AutoSizeUtil.appBuilder,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: HomePage(),
      ),
    )

```
3. 获取Size
```dart
AutoSizeUtil.getScreenSize()
//或者
MediaQuery.of(context).size
```
4. 直接按照设计图的尺寸写即可
```dart
Container(
    alignment: Alignment.center,
    height: 60,
    width :60,
    color: Colors.redAccent,
    child: Text(
      "直接按照设计图写尺寸",
    ),
  )
```
5. 获取真正的大小，比如 kToolbarHeight 是 56
```dart
// 一下方法获取到真正的大小，然后设置到 toolbar上
AutoSizeUtil.getRealSize(kToolbarHeight)
```
6. 切记
不能使用 window 获取 size 或者是 获取  MediaQuery<br>
~~window.physicalSize~~<br>
~~MediaQueryData.fromWindow(window)~~<br>

## 下一个目标
因为从根修改了屏幕宽高, 会引起一些系统Widget，有可能比正常的大或者小，比如toorbar的高度默认是 kToolbarHeight = 56.0，这样如果把宽度标准设置成112的话，那这kToolbarHeight就会显得很高，
还有 Text （bodyText2） 默认的是 14，也会非常大，所以 下一个计划是 还原一下默认的系统 widget。大家有好的建议 也可以提给我

## 原理
[掘进](https://juejin.cn/post/6996924953320751134)
## Thank
[FlutterTest](https://github.com/genius158/FlutterTest)






CHANGELOG.md
## 1.0.1
版本号兼容
## 0.1.3
修改 MediaQuery.of(context).padding 以及 viewPadding，适配 SafeArea
## 0.1.2
解决软键盘打开会有空隙，MediaQuery.of(context).viewInsets.bottom
## 0.1.1
- 添加获取真正的大小
- 添加是否随着系统文字大小改变而改变
## 0.0.1
初步完成适配情况


lib/temp.dart
import 'dart:async';
import 'dart:collection';
import 'dart:ui';

import 'package:flutter/gestures.dart';
import 'package:flutter/rendering.dart';
import 'package:flutter/widgets.dart';

import 'auto_size_util.dart';

class AutoSizeWidgetsFlutterBinding extends WidgetsFlutterBinding {
  static WidgetsBinding? ensureInitialized() {
    if (WidgetsBinding.instance == null) AutoSizeWidgetsFlutterBinding();
    return WidgetsBinding.instance;
  }

  @override
  ViewConfiguration createViewConfiguration() {
    var originalSize = window.physicalSize / window.devicePixelRatio;

    var originalWidth = originalSize.width;
    var originalHeight = originalSize.height;
    AutoSizeUtil.devicePixelRatio =
        window.physicalSize.width / AutoSizeUtil.screenWidth;

    return ViewConfiguration(
      size: Size(AutoSizeUtil.screenWidth,
          originalHeight / ((originalWidth / AutoSizeUtil.screenWidth))),
      // size: AutoSizeUtils.screenSize,
      devicePixelRatio: AutoSizeUtil.devicePixelRatio,
    );
  }

  /// 因为调整了 PixelRatio，所以对于 Event 事件，需要额外对事件的坐标进行对应比例的转换；
  @override
  void initInstances() {
    super.initInstances();
    window.onPointerDataPacket = _handlePointerDataPacket;
  }

  @override
  void unlocked() {
    super.unlocked();
    _flushPointerEventQueue();
  }

  final Queue<PointerEvent> _pendingPointerEvents = Queue<PointerEvent>();

  void _handlePointerDataPacket(PointerDataPacket packet) {
    // Event 事件，需要按比例转换尺寸
    _pendingPointerEvents.addAll(PointerEventConverter.expand(
        packet.data, AutoSizeUtil.devicePixelRatio));
    if (!locked) _flushPointerEventQueue();
  }

  @override
  void cancelPointer(int pointer) {
    if (_pendingPointerEvents.isEmpty && !locked)
      scheduleMicrotask(_flushPointerEventQueue);
    _pendingPointerEvents.addFirst(PointerCancelEvent(pointer: pointer));
  }

  void _flushPointerEventQueue() {
    assert(!locked);
    while (_pendingPointerEvents.isNotEmpty)
      _handlePointerEvent(_pendingPointerEvents.removeFirst());
  }

  final Map<int, HitTestResult> _hitTests = <int, HitTestResult>{};

  void _handlePointerEvent(PointerEvent event) {
    assert(!locked);
    HitTestResult result;
    if (event is PointerDownEvent) {
      assert(!_hitTests.containsKey(event.pointer));
      result = HitTestResult();
      hitTest(result, event.position);
      _hitTests[event.pointer] = result;
    } else if (event is PointerUpEvent || event is PointerCancelEvent) {
      result = _hitTests.remove(event.pointer)!;
    } else if (event.down) {
      result = _hitTests[event.pointer]!;
    } else {
      return; // We currently ignore add, remove, and hover move events.
    }
    if (result != null) dispatchEvent(event, result);
  }
}

/**
 * 提供 WidgetsFlutterBinding 的同名方法
 */
void ensureAutoSizeInitialized() {
  AutoSizeWidgetsFlutterBinding.ensureInitialized();
}

/**
 * 替代 runApp()
 */
void runAutoSizeApp(Widget app) {
  var runAppCallback = () {
    AutoSizeWidgetsFlutterBinding.ensureInitialized()!
      ..scheduleAttachRootWidget(app)
      ..scheduleWarmUpFrame();
  };
  // 屏幕尺寸初始化较晚，此操作可限制在尺寸赋值后继续操作
  if (window.physicalSize.isEmpty) {
    VoidCallback? oldMetricsChanged = window.onMetricsChanged;
    window.onMetricsChanged = () {
      if (!window.physicalSize.isEmpty) {
        if (oldMetricsChanged != null) {
          oldMetricsChanged();
        }
        window.onMetricsChanged = oldMetricsChanged;
        runAppCallback();
      }
    };
  } else {
    runAppCallback();
  }
}


lib/binding.dart
import 'dart:async';
import 'dart:collection';
import 'dart:ui';

import 'package:flutter/gestures.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

import 'auto_size_util.dart';


void runAutoApp(Widget app) {
  AutoWidgetsFlutterBinding.ensureInitialized()
    ..scheduleAttachRootWidget(app)
    ..scheduleWarmUpFrame();
}


class AutoWidgetsFlutterBinding extends WidgetsFlutterBinding {
  static WidgetsBinding ensureInitialized() {
    if (WidgetsBinding.instance == null) AutoWidgetsFlutterBinding();
    return WidgetsBinding.instance!;
  }

  @override
  ViewConfiguration createViewConfiguration() {
    return ViewConfiguration(
      size: AutoSizeUtil.getSize(),
      devicePixelRatio: AutoSizeUtil.getDevicePixelRatio(),
    );
  }

  /// 因为修改了 devicePixelRatio ，得适配点击事件  GestureBinding
  @override
  void initInstances() {
    super.initInstances();
    window.onPointerDataPacket = _handlePointerDataPacket;
  }

  @override
  void unlocked() {
    super.unlocked();
    _flushPointerEventQueue();
  }

  final Queue<PointerEvent> _pendingPointerEvents = Queue<PointerEvent>();

  void _handlePointerDataPacket(PointerDataPacket packet) {
    _pendingPointerEvents.addAll(PointerEventConverter.expand(
        packet.data,
        // 适配事件的转换比率,采用我们修改的
        AutoSizeUtil.getDevicePixelRatio()));
    if (!locked) _flushPointerEventQueue();
  }

  @override
  void cancelPointer(int pointer) {
    if (_pendingPointerEvents.isEmpty && !locked)
      scheduleMicrotask(_flushPointerEventQueue);
    _pendingPointerEvents.addFirst(PointerCancelEvent(pointer: pointer));
  }

  void _flushPointerEventQueue() {
    assert(!locked);
    while (_pendingPointerEvents.isNotEmpty)
      _handlePointerEvent(_pendingPointerEvents.removeFirst());
  }

  final Map<int, HitTestResult> _hitTests = <int, HitTestResult>{};

  void _handlePointerEvent(PointerEvent event) {
    assert(!locked);
    HitTestResult result;
    if (event is PointerDownEvent) {
      assert(!_hitTests.containsKey(event.pointer));
      result = HitTestResult();
      hitTest(result, event.position);
      _hitTests[event.pointer] = result;
      assert(() {
        if (debugPrintHitTestResults) debugPrint('$event: $result');
        return true;
      }());
    } else if (event is PointerUpEvent || event is PointerCancelEvent) {
      result = _hitTests.remove(event.pointer)!;
    } else if (event.down) {
      result = _hitTests[event.pointer]!;
    } else {
      return; // We currently ignore add, remove, and hover move events.
    }
    if (result != null) dispatchEvent(event, result);
  }
}


lib/auto_size_util.dart
import 'dart:ui';

import 'package:flutter/material.dart';

class AutoSizeUtil {
  static double _devicePixelRatio = 3.0;

  static double _screenWidth = 300;

  static double _screenHeight = 800;

  static double _screenStandard = 360;

  static Size _screenSize = Size.zero;

  static bool _autoTextSize = true;

  /// 如果是横屏 就以高度为基准
  /// 如果是竖屏 就以宽度为基准
  /// 是否随着系统的文字大小而改变，默认是改变
  static void setStandard(double standard, {bool isAutoTextSize = true}) {
    _screenStandard = standard;
    _autoTextSize = isAutoTextSize;
  }

  /// 根据设置 的 宽度 来得到 devicePixelRatio
  static double getDevicePixelRatio() {
    var originalSize = window.physicalSize / window.devicePixelRatio;
    var originalWidth = originalSize.width;
    var originalHeight = originalSize.height;
    if (originalHeight > originalWidth) {
      // 竖屏
      _devicePixelRatio =
          window.physicalSize.width / AutoSizeUtil._screenStandard;
    } else {
      // 横屏
      _devicePixelRatio =
          window.physicalSize.height / AutoSizeUtil._screenStandard;
    }
    return _devicePixelRatio;
  }

  /// 根据设置的宽度，来得到对应的高度
  static Size getSize() {
    // 如果是横屏就已宽度为基准

    var originalSize = window.physicalSize / window.devicePixelRatio;
    var originalWidth = originalSize.width;
    var originalHeight = originalSize.height;
    if (originalHeight > originalWidth) {
      // 竖屏
      _screenHeight = window.physicalSize.height / getDevicePixelRatio();
      _screenWidth = _screenStandard;
      _screenSize = Size(_screenStandard, _screenHeight);
      return _screenSize;
    } else {
      // 横屏
      _screenWidth = window.physicalSize.width / getDevicePixelRatio();
      _screenHeight = _screenStandard;
      _screenSize = Size(_screenWidth, _screenStandard);
      return _screenSize;
    }
  }

  static Size getScreenSize() {
    return _screenSize;
  }

  static Widget appBuilder(BuildContext context, Widget? widget) {
    EdgeInsets viewInsets =  MediaQuery.of(context).viewInsets;
    EdgeInsets pading =  MediaQuery.of(context).padding;
    EdgeInsets viewPadding =  MediaQuery.of(context).viewPadding;
    var adapterEdge=  EdgeInsets.fromLTRB(viewInsets.left, viewInsets.top, viewInsets.right,getRealSize(viewInsets.bottom));
    var adapterPadding=  EdgeInsets.fromLTRB(pading.left, getRealSize(pading.top), pading.right,getRealSize(pading.bottom));
    var adapterViewPadding=  EdgeInsets.fromLTRB(viewPadding.left, getRealSize(viewPadding.top), viewPadding.right,getRealSize(viewPadding.bottom));

    return MediaQuery(
      // 这里如果设置 textScaleFactor = 1.0 ，就不会随着系统字体大小去改变了
      data: MediaQuery.of(context).copyWith(
          size: Size(AutoSizeUtil._screenWidth, AutoSizeUtil._screenHeight),
          devicePixelRatio: AutoSizeUtil._devicePixelRatio,
          textScaleFactor:
              _autoTextSize ? MediaQuery.of(context).textScaleFactor : 1.0,viewInsets:adapterEdge,padding: adapterPadding,viewPadding: adapterViewPadding),
      child: _adapterTheme(context, widget),
    );
  }

  /// 获取真正的大小，比如 kToolbarHeight kBottomNavigationBarHeight
  static double getRealSize(double size) {
    return size / (_devicePixelRatio / window.devicePixelRatio);
  }

  static _adapterTheme(BuildContext context, Widget? widget) {

    return Theme(
      data: Theme.of(context).copyWith(),
      child: widget!,
    );
  }
}


example/lib/main.dart
import 'dart:ui';

import 'package:flutter/material.dart';
import 'package:flutter_autosize_screen/auto_size_util.dart';
import 'package:flutter_autosize_screen/binding.dart';

void main() {
  AutoSizeUtil.setStandard(360,isAutoTextSize: true);
  runAutoApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Autosize Demo',
      builder: AutoSizeUtil.appBuilder,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: HomePage(),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  State<HomePage> createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  TextStyle _style = TextStyle(color: Colors.white);
  GlobalKey _keyGreen = GlobalKey();
  GlobalKey _keyBlue = GlobalKey();

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance!.addPostFrameCallback((timeStamp) {
      var renderBox = _keyGreen.currentContext!.findRenderObject()!.paintBounds;
      var sizeGreen = renderBox.size;
      print("${sizeGreen.width} ----- ${sizeGreen.height}");

      var renderBlu = _keyBlue.currentContext!.findRenderObject()!.paintBounds;
      var sizeBlue = renderBlu.size;
      print("${sizeBlue.width} ----- ${sizeBlue.height}");
      print("${AutoSizeUtil.getScreenSize()}");
    });
  }

  @override
  Widget build(BuildContext context) {
    var originalSize = window.physicalSize / window.devicePixelRatio;
    var nowDevicePixelRatio = MediaQuery.of(context).devicePixelRatio;

    return Scaffold(
      appBar: AppBar(
        title: Text("Autosize Demo"),
      ),
      body: SingleChildScrollView(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          children: [
            Row(
              children: [
                Expanded(
                  child: Container(
                    alignment: Alignment.center,
                    key: _keyGreen,
                    height: 60,
                    color: Colors.redAccent,
                    child: Text(
                      "使用Expanded平分屏幕",
                      style: _style,
                    ),
                  ),
                ),
                Expanded(
                  child: Container(
                    alignment: Alignment.center,
                    height: 60,
                    color: Colors.blue,
                    child: Text(
                      "使用Expanded平分屏幕",
                      style: _style,
                    ),
                  ),
                ),
              ],
            ),
            Row(
              children: [
                Container(
                  alignment: Alignment.center,
                  key: _keyBlue,
                  width: 180,
                  height: 60,
                  color: Colors.teal,
                  child: Text(
                    "宽度写的是 180",
                    style: _style,
                  ),
                ),
                Container(
                  alignment: Alignment.center,
                  width: 180,
                  height: 60,
                  color: Colors.grey,
                  child: Text(
                    "宽度写的是 180",
                    style: _style,
                  ),
                ),
              ],
            ),
            Container(
              alignment: Alignment.center,
              width: 360,
              height: 60,
              color: Colors.purple,
              child: Text(
                "宽度写的是 360",
                style: _style,
              ),
            ),
            SizedBox(
              height: 50,
            ),
            Text("原始的 size: $originalSize "),
            Text("原始的 分辨率: ${window.physicalSize} "),
            Text("原始的 devicePixelRatio: ${window.devicePixelRatio} "),
            Container(
              margin: EdgeInsets.symmetric(vertical: 20),
              alignment: Alignment.center,
              width: 360,
              height: 10,
              color: Colors.grey,
            ),
            Text("更改后 size: ${MediaQuery.of(context).size}  "),
            Text("更改后 devicePixelRatio: $nowDevicePixelRatio"),
          ],
        ),
      ),
    );
  }
}
